baseapp/abci.go

```diff
@@ -248,11 +246,26 @@ func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBloc
 // Ref: https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-060-abci-1.0.md
 // Ref: https://github.com/tendermint/tendermint/blob/main/spec/abci/abci%2B%2B_basic_concepts.md
 func (app *BaseApp) PrepareProposal(req abci.RequestPrepareProposal) (resp abci.ResponsePrepareProposal) {
-	ctx := app.getContextForTx(runTxPrepareProposal, []byte{})
 	if app.prepareProposal == nil {
 		panic("PrepareProposal method not set")
 	}
 
+	if app.deliverState == nil {
+		initHeader := tmproto.Header{
+			Time:            req.Time,
+			Height:          req.Height,
+			ProposerAddress: req.ProposerAddress,
+		}
+		app.setState(runTxModeDeliver, initHeader)
```

Comment:
Let's update `setState` to also accept `runTxPrepareProposal` for deliver state.

i.e. `case runTxModeDeliver, runTxPrepareProposal: ...`